File: src/main/java/com/tradingsupervisor/viewmodel/LocationLiveData.java
package com.tradingsupervisor.viewmodel;

import android.annotation.SuppressLint;
import android.arch.lifecycle.LiveData;
import android.content.Context;
import android.location.Location;
import android.util.Log;


import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;

public class LocationLiveData extends LiveData<Location> {

    private FusedLocationProviderClient locationClient;
    private LocationRequest locationRequest;

    @SuppressLint("MissingPermission")
    public LocationLiveData(Context appContext) {
        locationClient = LocationServices.getFusedLocationProviderClient(appContext);
        locationRequest = new LocationRequest();
        locationRequest.setInterval(50000);
        locationRequest.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY); //PRIORITY_HIGH_ACCURACY
    }


    private LocationCallback locationCallback = new LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            setValue(locationResult.getLastLocation());
            Log.d("upd", locationResult.getLastLocation().toString());
        }
    };

    @SuppressLint("MissingPermission")
    @Override
    protected void onActive() {
        super.onActive();
        locationClient.requestLocationUpdates(locationRequest, locationCallback, null);
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        if (locationCallback != null)
            locationClient.removeLocationUpdates(locationCallback);
    }
}

File: app/src/main/java/com/tradingsupervisor/viewmodel/LocationLiveData.kt
package com.tradingsupervisor.viewmodel

import android.annotation.SuppressLint
import android.content.Context
import android.location.Location
import android.os.HandlerThread
import android.util.Log
import androidx.lifecycle.LiveData
import com.google.android.gms.location.*

class LocationLiveData private constructor(appContext: Context) : LiveData<Location>() {
    private val locationClient = LocationServices.getFusedLocationProviderClient(appContext)
    private var handlerThread: HandlerThread? = null
    private val locationRequest: LocationRequest = LocationRequest().apply {
        interval = (10 * 1000).toLong() //10 sec
        fastestInterval = (10 * 1000).toLong()
        priority = LocationRequest.PRIORITY_HIGH_ACCURACY
    }

    private val locationCallback: LocationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult?) {
            locationResult ?: return
            super.onLocationResult(locationResult)
            val location = locationResult.lastLocation
            //val lat = location.latitude.toString()
            //val lon = location.longitude.toString()
            //Log.d("myupd", "new location    " + this.hashCode() + "   " + lat + " : " + lon);
            postValue(location)
        }
    }

    @SuppressLint("MissingPermission")
    override fun onActive() {
        super.onActive()
        handlerThread = HandlerThread("locationHandlerThread")
        handlerThread!!.start()
        locationClient.requestLocationUpdates(locationRequest, locationCallback, handlerThread!!.looper)
    }

    override fun onInactive() {
        super.onInactive()
        locationClient.removeLocationUpdates(locationCallback)
        handlerThread!!.quitSafely()
        //setValue(new Location(""));
    }

    companion object {
        fun getInstance(appContext: Context): LocationLiveData {
            return LocationLiveData(appContext)
        }
    }
}